flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java

```diff
@@ -57,6 +57,9 @@
   private static final String ICEBERG_STREAM_WRITER_NAME = IcebergStreamWriter.class.getSimpleName();
   private static final String ICEBERG_FILES_COMMITTER_NAME = IcebergFilesCommitter.class.getSimpleName();
 
+  public static final String FLINK_ICEBERG_SINK_FLUSHINTERVAL = "flink.iceberg.sink.flushinterval";
```

Comment:
Can we be sure to document (preferably both in the code as well as in the docs on the website) that this configuration is only used if checkpointing is disabled? That's an important point but it seems easy to miss without looking through the code.

Given the name, as an average developer who uses Flink but possibly does not know all of the details of its usage and hopefully at least scans the documentations for various configurations, I imagine I would personally assume that this defines the flush interval.... full stop. Not the flush interval if and only if checkpointing is disabled. To me, I would think that the average Flink developer might liken this to Kafka's `linger.ms` producer configuration in a sense. Maybe this is just from my background of onboarding developers who aren't always well versed with "big data" to Flink.

Maybe `flink.iceberg.sink.flushinterval.when.committing.disabled` (or possibly using dashes for the last part of the name) or something else might be more clear? Seems a little verbose for me, but as someone who oversees a variety of developers with differing backgrounds that use Flink, this configuration property seems likely to cause confusion without further clarity.

I would prefer that we change the name so that reading a comment in the docs isn't required, but I'm open to both approaches.

Does anybody else have an opinion on this?

Reply:
Thanks kbendick, I think this configuration should be "flink.iceberg.sink.commit.interval" rather than flush interval. And we can add some comments or document to clarify this, what do you think ?

Reply:
Hi @simonsssu. Thanks for the quick response.

To me, the new name suggested is potentially more confusing, as it introduces the term `commit interval` into the configuration for the `IcebergFileCommitter`, but it only covers a small (arguably edge) case for jobs that do not checkpoint. I think, for many Flink users, seeing the configuration `flink.iceberg.sink.commit.interval` will still very much lead them to believe that this is the main parameter to tune regardless of checkpoints. But my understanding is that we rely on:

1. Checkpoints to know when to commit.
2. This configuration parameter to commit when checkpointing is not enabled.

In Flink, when a checkpoint happens, there's possibly several commits, typically the last one being committing the id of the last checkpoint to zookeeper in a zookeeper based HA set up with checkpointing enabled, which is by far the most common setup amongst my Flink users (I have one user with only one job that does not checkpoint).

To show an example of where `commit` is already a preexisting term with ties to checkpoints in Flink, under the documentation for [High Availability in the Flink Configuration Docs](https://ci.apache.org/projects/flink/flink-docs-stable/ops/config.html#high-availability) the following sentence also uses commit, to refer to when the jobmanager commits metadata to zookeeper.

```
For the JobManager itself to recover consistently, an external service must store a minimal amount of recovery 
metadata (like “ID of last committed checkpoint”)...
```

I'd still really prefer to see something in the configuration name that indicates this configuration interval is _only when not checkpointing_. I worry / have observed that people frequently don't read the docs as much as they should and then encounter issues. However, I do recognize that we probably don't want to make the configuration names really verbose.

What about `flink.iceberg.non-checkpointing-job.sink-flush-interval` or to be more inline with flink's naming pattern for  configurations, `flink.iceberg.non.checkpointing.job.sink-commit-interval`? This is more inline with flink's own cluster level parameter to control the checkpoint interval, [execution.checkpointing.interval](https://ci.apache.org/projects/flink/flink-docs-stable/ops/config.html#execution-checkpointing-interval). Additionally the usage of interval here corresponds nicely to the Flink configuration parameter's description for controlling the maximum interval at which network buffers should be flushed, [execution.buffer-timeout](https://ci.apache.org/projects/flink/flink-docs-stable/ops/config.html#execution-buffer-timeout). So possibly a name that mirrors the `execution.buffer-timeout`, which controls the maximum time frequency for the flushing of (task to task / shuffling) network buffers.

Another name could be `flink.iceberg.non-checkpointing-sink-buffer-flush-interval`.

I think of all of the options I mentioned above, the one that I am most leaning towards that I think most clearly explains the paremeter is `flink.iceberg.non.checkpointing.job.sink-flush-interval` or `flink.iceberg.sink-flush-interval-for-non-checkpointing-jobs`. I think the second is the most clear as it has `sink` right after `iceberg` and has `for-non-checkpointing-jobs`. I'd still like to see it documented in the code via a comment and even in the Flink docs on the website.

What do you think of these names @simonsssu? I'd love to hear any suggestions for alternative config names, especially ones that emphasize that this is for non-checkpointing jobs, though I will admit that I think `flink.iceberg.sink.commit.interval` adds further confusion. If for example, I were a new dev in an organization that used flink, and I was writing my first job and copy pasting from another job to create a skeleton, if I saw `flink.iceberg.sink.commit.interval` I would assume that this was the one and only parameter that controlled when files are committed. But as I understand it, files are committed on job checkpoint and then we'll fall back to this parameter if the job doesn't checkpoint.

Reply:
Oh and I have another idea to help ensure that developers follow the happy path when configuring their jobs.

We know in the `FlinkIcebergCommitter#open` method whether or not this job is checkpointing. We set the parameter (whatever it winds up being called) in the operators constructor.

After checking for `isCheckpointingEnabled`, could we throw an exception if a value is set for the new parameter? It would require to not eagerly set the default value in the constructor, but instead set it lazily in `open` as well. This is beneficial given that the usage of `FlinkConfiguration.getLong(String key, long defaultValue)` is deprecated anyway.

Admittedly, my proposed idea still relies on the deprecated method until such a time that we can define our own `ConfigOption` and then use one of the two recommended replacement methods: 

-  `public <T> Optional<T> getOptional(ConfigOption<T> option)`
- `public long getLong(ConfigOption<Long> configOption, long overrideDefault)`

```java
  @Override
  public void open() throws Exception {
    super.open();
    final boolean isCheckpointEnabled = getRuntimeContext().isCheckpointingEnabled();
    final boolean isCheckpointDisabled = !isCheckpointEnabled;
    // It is considered an error for users to set FlinkSink.FLINK_ICEBERG_SINK_FLUSHINTERVAL
    // when checkpointing is enabled, as they likely intended for data flushing to occur at
    // the FLINK_ICEBERG_SINK_FLUSHINTERVAL, but it will instead occur according to the
    // configured checkpointing interval.
    //
    // Note (that shouldn't go in the checked in code) that I'm using the existing config constant name,
    // though I'd still love to see it updated and then update all references to the new name!
    //
    // Please import org.apache.iceberg.exceptions.ValidationException and then use ValidationException.check
    // I used the fully qualified class name to show you where this can be found (please also do not include this
    // comment in the code).
    org.apache.iceberg.exceptions.ValidationException.check(
        isCheckpointEnabled && this.flinkConf.contains(FlinkSink.FLINK_ICEBERG_SINK_FLUSHINTERVAL),
        "Cannot set %s when checkpointing is enabled. It is unlikely this is what is intended. Data will be buffered until the next checkpoint, and then flushed and committed. To control the rate of flushing iceberg , change the checkpointing interval.",
        FlinkSink.FLINK_ICEBERG_SINK_FLUSHINTERVAL
    );
    // If we don't enable checkpoint, we will use processingTimeSerice to do commit,
    if (isCheckpointDisabled) {
      this.flushCommitInterval = this.flinkConf.getLong(
                                                      FlinkSink.FLINK_ICEBERG_SINK_FLUSHINTERVAL,
                                                      FlinkSink.DEFAULT_FLINK_ICEBERG_SINK_FLUSHINTERVAL);
      ProcessingTimeService processingTimeService = getRuntimeContext().getProcessingTimeService();
      final long currentTimestamp = processingTimeService.getCurrentProcessingTime();
      processingTimeService.registerTimer(currentTimestamp + flushCommitInterval, this);

      // Log to the user what value in processing time will be used for flush / commit, and where it was
      // derived from, as most likely users have just forgotten to enable checkpointing.
      final String logIntervalFmtStr = 
        "Checkpointing is disabled. The IcebergFilesCommitter will flush data files and commit every %d millis as determined by %s.";
      final String commitIntervalSource = this.flinkConf.contains(FlinkSink.FLINK_ICEBERG_SINK_FLUSHINTERVAL)
                                                                     ? FlinkSink.FLINK_ICEBERG_SINK_FLUSHINTERVAL
                                                                     : FlinkSink.DEFAULT_FLINK_ICEBERG_SINK_FLUSHINTERVAL;
        LOG.info(logIntervalFmtStr, this.flushCommitInterval, commitIntervalSource);
    }
```

I think that we should treat jobs that have checkpointing enabled with a specified flink iceberg sink flush / commit interval (whatever we wind up naming it) as an error. If such a time comes that we allow users to specify the commit interval for the iceberg sink independent of the checkpointing interval, then we can stop treating it as an error

I imagine for most users, winding up in a situation where they're not checkpointing in a very stateful job and implicitly using the default flush interval is probably not what they intended and indicates a bug on their end - e.g. they might not know that checkpointing is not enabled (as it's pretty rare to disable it in production for jobs that depend on state so much). Thus, I think the log will help developers determine if they need to make changes to their configuration.

Reply:
> Hi @simonsssu. Thanks for the quick response.
> 
> To me, the new name suggested is potentially more confusing, as it introduces the term `commit interval` into the configuration for the `IcebergFileCommitter`, but it only covers a small (arguably edge) case for jobs that do not checkpoint. I think, for many Flink users, seeing the configuration `flink.iceberg.sink.commit.interval` will still very much lead them to believe that this is the main parameter to tune regardless of checkpoints. But my understanding is that we rely on:
> 
> 1. Checkpoints to know when to commit.
> 2. This configuration parameter to commit when checkpointing is not enabled.
> 
> In Flink, when a checkpoint happens, there's possibly several commits, typically the last one being committing the id of the last checkpoint to zookeeper in a zookeeper based HA set up with checkpointing enabled, which is by far the most common setup amongst my Flink users (I have one user with only one job that does not checkpoint).
> 
> To show an example of where `commit` is already a preexisting term with ties to checkpoints in Flink, under the documentation for [High Availability in the Flink Configuration Docs](https://ci.apache.org/projects/flink/flink-docs-stable/ops/config.html#high-availability) the following sentence also uses commit, to refer to when the jobmanager commits metadata to zookeeper.
> 
> ```
> For the JobManager itself to recover consistently, an external service must store a minimal 
> amount of recovery metadata (like “ID of last committed checkpoint”)
> ```
> 
> I'd still really prefer to see something in the configuration name that indicates this configuration interval is _only when not checkpointing_. I worry / have observed that people frequently don't read the docs as much as they should and then encounter issues. However, I do recognize that we probably don't want to make the configuration names really verbose.
> 
> What about `flink.iceberg.non-checkpointing-job.sink-flush-interval` or to be more inline with flink's naming pattern for configurations, `flink.iceberg.non.checkpointing.job.sink-commit-interval`? This is more inline with flink's own cluster level parameter to control the checkpoint interval, [execution.checkpointing.interval](https://ci.apache.org/projects/flink/flink-docs-stable/ops/config.html#execution-checkpointing-interval). Additionally the usage of interval here corresponds nicely to the Flink configuration parameter's description for controlling the maximum interval at which network buffers should be flushed, [execution.buffer-timeout](https://ci.apache.org/projects/flink/flink-docs-stable/ops/config.html#execution-buffer-timeout). So possibly a name that mirrors the `execution.buffer-timeout`, which controls the maximum time frequency for the flushing of (task to task / shuffling) network buffers.
> 
> Another name could be `flink.iceberg.non-checkpointing-sink-buffer-flush-interval`.
> 
> I think of all of the options I mentioned above, the one that I am most leaning towards that I think most clearly explains the paremeter is `flink.iceberg.non.checkpointing.job.sink-flush-interval` or `flink.iceberg.sink-flush-interval-for-non-checkpointing-jobs`. I think the second is the most clear as it has `sink` right after `iceberg` and has `for-non-checkpointing-jobs`. I'd still like to see it documented in the code via a comment and even in the Flink docs on the website.
> 
> What do you think of these names @simonsssu? I'd love to hear any suggestions for alternative config names, especially ones that emphasize that this is for non-checkpointing jobs, though I will admit that I think `flink.iceberg.sink.commit.interval` adds further confusion. If for example, I were a new dev in an organization that used flink, and I was writing my first job and copy pasting from another job to create a skeleton, if I saw `flink.iceberg.sink.commit.interval` I would assume that this was the one and only parameter that controlled when files are committed. But as I understand it, files are committed on job checkpoint and then we'll fall back to this parameter if the job doesn't checkpoint.

Hi Kyle,  Apparently,  Word by word configuration name will not make confused. But I think it's not elegant if we defined a too long configuration name. So I think we can use  " flink.iceberg.sink.non-checkpoint-commit-interval" and add documents to tell users about the details. Just add non-checkpoint in my previous defined.

Reply:
Hi Kyle, Apparently, Word by word configuration name will not make confused. But I think it's not elegant if we defined a too long configuration name. So I think we can use " flink.iceberg.sink.non-checkpoint-commit-interval" and add documents to tell users about the details. Just add non-checkpoint in my previous defined.

Reply:
I agree, we can do some check in open() method to make sure that the developers will follow the right behaviors.